Sources/NIO/ByteBuffer-views.swift

```diff
@@ -16,48 +16,161 @@
+            guard position >= self._range.lowerBound && position < self._range.upperBound else {
+                preconditionFailure("index \(position) out of range")
+            }
+            return self._buffer.getInteger(at: position)! // range check above
+        }
+        set {
+            guard position >= self._range.lowerBound && position < self._range.upperBound else {
+                preconditionFailure("index \(position) out of range")
+            }
+            self._buffer.setInteger(newValue, at: position)
         }
-        return self.buffer.getInteger(at: position)! // range check above
     }
 
     public subscript(range: Range<Index>) -> ByteBufferView {
-        return ByteBufferView(buffer: self.buffer, range: range)
+        get {
+            return ByteBufferView(buffer: self._buffer, range: range)
+        }
+        set {
+            self.replaceSubrange(range, with: newValue)
+        }
+    }
+
+    @inlinable
+    public mutating func replaceSubrange<C: Collection>(_ subrange: Range<Index>, with newElements: C) where ByteBufferView.Element == C.Element {
+        precondition(subrange.lowerBound >= self.startIndex && subrange.upperBound <= self.endIndex,
+                     "subrange out of bounds")
+
+        if newElements.count <= subrange.count {
+            // Copy the bytes over into the subrange.
+            self._copyBytes(to: subrange, from: newElements)
+
+            // `newElements` is shorter than subrange, so we now have unwanted bytes between the
+            // copied bytes and the end of the subrange. Let's move all of the bytes after the
+            // subrange taking care to update the buffer's writer index as well.
+            if newElements.count < subrange.count {
+                let oldWriterIndex = self._buffer.writerIndex
+                let removedBytes = subrange.count - newElements.count
+                self._buffer.moveWriterIndex(to: oldWriterIndex - removedBytes)
+                self._range = self._range.dropLast(removedBytes)
+
+                self._buffer.withVeryUnsafeMutableBytes { ptr in
+                    let endOfRangeToReplace = subrange.upperBound
+                    let srcPtr = ptr.baseAddress!.advanced(by: endOfRangeToReplace)
+
+                    let endOfReplacedRange = subrange.lowerBound.advanced(by: newElements.count)
+                    let dstPtr = ptr.baseAddress!.advanced(by: endOfReplacedRange)
+
+                    let numberOfBytesToMove = endOfRangeToReplace.distance(to: oldWriterIndex)
+                    memmove(dstPtr, srcPtr, numberOfBytesToMove)
+                }
+            }
+        } else {
+            // We have more elements to write in the subrange than the subrange currently has space
+            // for: widen the capacity of the underlying buffer.
+            let additionalByteCount = newElements.count - subrange.count
+            self._buffer.reserveCapacity(self._buffer.readableBytes + additionalByteCount)
```

Comment:
this isn't quite right I think, this should be `self._buffer.writerIndex + additionalByteCount`.